### DISCUSSION TITLE
Replacement IMU

### DISCUSSION THREAD

**Discussion by DrewBrandt** [Terrapin-Rocket-Team/SRAD-Avionics#106] [Ideas]
The BNO055 apparently has a lot of issues that we don't need to deal with. Chief among those include a bad I2C implementation, sub-par sensor calibration, and outdated firmware.

Basic internet searches suggest that any modern sensor should work better, so here I'll catalogue what I find along with its pros and cons. Feel free to find a different sensor (or group of sensors) and add them below so we can compare them before coming to a final decision.

All pros and cons will be compared to the baseline of "BNO055 features."

[An interesting link on various gyroscopes in some of these sensors](https://learn.adafruit.com/comparing-gyroscope-datasheets)

**Comment by DrewBrandt** [comment]
# BNO085

### Pros:
- Automatic calibration storage
- Newer firmware

### Cons:
- Same old hardware in a fancy new box
- Has same timing violation issues that BNO055 has

**Reply by varun-un** [reply]
Is 16g mode a problem here as well or should this be fine?

**Reply by DrewBrandt** [reply]
To be honest, I don't know why 16g mode didn't work on the BNO and whether it was a software bug or a hardware bug. So I don't know if would be an issue again if we moved to the 085.

**Comment by DrewBrandt** [comment]
# Literally any other IMU (LSM9DS1/ICM20948/etc)

### Pros:
- Cheaper
- No I2C Timing violations
- No faulty calibration or fusion firmware
### Cons:
- No built-in "Absolute" sensor orientation. Math must be done on MCU. (not a con tho if the BNO is faulty, or if we don't really care about just doing that math. It's not that hard)

**Reply by varun-un** [reply]
I feel like biggest downside of absolute sensor orientation isn't necessarily the absolute part, since that frame transformation math shouldn't be that painful, but the hard part is making the internal filtering through like an IIR or something so that the data we read out is good

**Reply by varun-un** [reply]
Also, did you find the accuracy or error bars of these IMUs, because for like the 72LSM9DS1 for example; it seems to imply that the error comes from the bit depth of the ADC when talking about symbols like `LA_So` and `M_GN` in the [datasheet](https://www.digikey.de/htmldatasheets/production/1639232/0/0/1/lsm9ds1-datasheet.html#pf11), unless I just missed it or it's a different symbol in that spec table

**Reply by DrewBrandt** [reply]
To me, for the the LA_So field, it looks like it's saying that it's minimum detectable change is 732 micro gs, which it is so far beyond our requirements that it really shouldn't be an issue. Sensor noise will be like 2 magnitudes greater than that amount, so I don't think we should worry.

**Reply by DrewBrandt** [reply]
# LSM9DS1

### Features:

- +- 16g
- LGA package
- 9DOF


### Notes:
- 0.7mg/LSB sensititvity in 16g mode
- +- 90 mg typical 0 offset in 8g mode
- $30 on DigiKey

**Reply by DrewBrandt** [reply]
# ICM 20948

### Features:
- +- 16g
- On-Chip Programmable filters?
- QFN Package
- 9DOF

### Notes:
- $8.64 on DigiKey
- 0.488mg/LSB sensitivity at 16g mode
- This image
![image](https://github.com/user-attachments/assets/95dbf618-d9eb-4ddc-a5de-1db996dbcb8b)

**Comment by varun-un** [comment]
# [BMI270](https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bmi270-ds000.pdf)

### Pros
- +/- 0.4% sensitivity
- Configurable to +/- 16g ranges
- 2 times the sensitivity of the BNO055 (could also be a con due to impulse shock events)
- RMS noise allegedly <5mg
### Cons
- Apparently loading software on startup takes time (not preloaded)
- Higher 0g offset/drift than the BNO055
- Open loop

**Comment by varun-un** [comment]
# [ADIS16467](https://www.analog.com/media/en/technical-documentation/data-sheets/adis16467.pdf)

### Pros
- Very precise 
- 13 μg in-run bias stability
- In flight bias correction and stability filtering
- +/- 40g range
- 2000 g mechanical shock survivability
- Internal filtering outputs velocity
- QFN chip
- What USC RPL used for their spaceshot
### Cons
- $650 plus around $100 in additional components
- Less SRAD

**Comment by varun-un** [comment]
# [BMI088](https://www.bosch-sensortec.com/products/motion-sensors/imus/bmi088/)

### Pros
- cheaper than BMI270
- Closed loop
- +/- 24g mode
### Cons
- Slightly less accurate than BMI270
- 6 DoF

**Comment by varun-un** [comment]
# [ASM330LHB](https://www.st.com/en/mems-and-sensors/asm330lhb.html)

### Pros
- Less noise and theoretically more accurate than BMI270
- Machine Learning powered (?)
- Android Auto™ compliant and CarPlay® compliant
### Cons
- 6 DoF
- Probably manual fusion and global calcs

**Comment by DrewBrandt** [comment]
# LSM6DSO32

### Features:

- 6DOF (no magnetometer)
- +- 32g range

### Notes:

- 0.976 mg/LSB sensitivity in 32g mode
- This image
![image](https://github.com/user-attachments/assets/207deeff-e959-421a-95be-d9abdb42d4d3)
